services/provider/server/server.go

```diff
@@ -901,6 +922,28 @@ func (s *OCSProviderServer) getOCSSubscriptionChannel(ctx context.Context) (stri
 	return subscription.Spec.Channel, nil
 }
 
+func (s *OCSProviderServer) getStorageCluster(ctx context.Context) (*ocsv1.StorageCluster, error) {
+	storageClusterList := &ocsv1.StorageClusterList{}
+	if err := s.client.List(ctx, storageClusterList, client.InNamespace(s.namespace)); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to list storage clusters: %v", err)
+	}
+	// Filter out external storage clusters if any
+	var storageClusters []ocsv1.StorageCluster
+	for _, cluster := range storageClusterList.Items {
+		if !cluster.Spec.ExternalStorage.Enable {
+			storageClusters = append(storageClusters, cluster)
+		}
+	}
+
+	if len(storageClusters) > 1 {
+		return nil, status.Errorf(codes.Internal, "found more than one storage cluster")
+	}
+	if len(storageClusters) == 0 {
+		return nil, status.Errorf(codes.NotFound, "no storage cluster found")
+	}
+	return &storageClusters[0], nil
```

Comment:
1. Don't use the `range` value capture, it will create an in-memory copy of the storage cluster - It is a very big structure

2. There is no need to create a new slice and start appending to it

```suggestion
	var found *ocsv1.StorageCluster = nil
	for i := range storageClusterList.Items {
		cluster := &storageClusterList.Items[i]
		if cluster.Spec.ExternalStorage.Enable {
			continue
		} 
		if found != nil {
			return nil, status.Errorf(codes.Internal, "found more than one storage cluster")
		}
		found = cluster
	}
	if found == nil {
		return nil, status.Errorf(codes.NotFound, "no storage cluster found")
	}
	return found
```

Reply:
Thanks for the detailed skeleton of the function, I have updated the function accordingly.